README.md
[![Build Status](https://travis-ci.org/cassproject/cass-vlrc.svg?branch=master)](https://travis-ci.org/cassproject/cass-vlrc)
# cass-vlrc
Competency and Skills Service -- Virtual Learning Resource Center

Claim competence for yourself or others, associate resources with competencies, and learn in a self-directed and competency-based example learning system.

# Features
 * View competency frameworks and see how many resources are associated with each competency.
 * Skills Inventory - Self-assert competencies in a framework for viewing in a profile or learning.
 * Resource Alignment - Associate resources with competencies, rate and view other people's resources.
 * Contacts and Peers - Find peers and make assertions about their competence as well!
 * Customizable - Change the default behavior of the cass-vlrc via URL parameters.
 * Unbranded and open source - Carries no branding information, allowing you to use it easily and without concern.

# Try it out
The (unbranded) cass-vlrc is available for use at https://vlrc.cassproject.org/

# Integration
Integrate the cass-vlrc into your website by nesting it into an iframe on your website.

```html
<iframe id="iframe" width="100%" height="800" frameborder="0"></iframe>
<script>
    document.getElementsByTagName("iframe")[0].setAttribute("src", "https://vlrc.cassproject.org/");
</script>
```

# URL Parameters
The cass-vlrc has many modes and options that can be enabled or disabled based on desired functionality. These options are passed as URL parameters.

To ensure the iframe is passing the correct origin, pass the origin into the iframe using a URL parameter, `origin=window.location.origin`.

## Setting the server
By default, the cass-vlrc operates on the dev.cassproject.org CASS instance. This should instead be your server. To do this, use `server=protocol://cass.yourserver.com`.

    ex:
    https://vlrc.cassproject.org/index.html?server=https://sandbox.cassproject.org

## Jump to a specific framework
If you have a specific framework you are directing your users to edit or view, you can force the cass-editor to only show that framework by using `frameworkId=<url identifier of framework>`

    ex:
    https://vlrc.cassproject.org/index.html?frameworkId=https://dev.cassproject.org/api/data/schema.cassproject.org.0.3.Framework/e17ad428-e27e-4265-b2ec-782fe7cd81e1
    
## Generate and manage a local user.
cass-vlrc will, by default, create and manage a local user that owns the assertions and resources the user provides. The user credentials will be stored in HTML5 LocalStorage and unique per individual.
  
  Note: This method is risky, as a user that clears their localstorage will lose access to their frameworks and competencies.
  
# Post Messages
CASS can operate using credentials and signals provided by the parent site. These credentials are passed into the iframe via a message, in order to avoid transmitting them over HTTP. Wait for a `{message:"waiting"}` message to be posted from the iframe.
```javascript
$("iframe").ready(function () {
    $(window).on("message", function (event) {
        if (event.originalEvent.data.message == "waiting") {
            //Initialize identity
        }
    });
});
```
## Operate using a provided set of user credentials.
CASS can operate using credentials provided by the parent site. These credentials are passed into the iframe via a message, in order to avoid transmitting them over HTTP. Use `user=wait`, wait for a `{message:"waiting"}` message, and pass a RSA-OAEP 2048 bit PEM private key in via:

    document.getElementById("iframe").contentWindow.postMessage(
      JSON.stringify({action:"identity",identity:privateKey}),
      window.location.origin
    );
    

CHANGELOG.md
# Change Log

## [1.2.8](https://github.com/cassproject/cass-vlrc/tree/1.2.8) (2019-05-10)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.2.7...1.2.8)

## [1.2.7](https://github.com/cassproject/cass-vlrc/tree/1.2.7) (2019-04-26)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.2.6...1.2.7)

## [1.2.6](https://github.com/cassproject/cass-vlrc/tree/1.2.6) (2019-04-24)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.2.5...1.2.6)

## [1.2.5](https://github.com/cassproject/cass-vlrc/tree/1.2.5) (2019-04-24)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.2.4...1.2.5)

## [1.2.4](https://github.com/cassproject/cass-vlrc/tree/1.2.4) (2019-04-19)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.2.3...1.2.4)

**Closed issues:**

-  Find better way of incorporating "can/cant/computed state" into bar. [\#33](https://github.com/cassproject/cass-vlrc/issues/33)

## [1.2.3](https://github.com/cassproject/cass-vlrc/tree/1.2.3) (2019-04-15)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.2.2...1.2.3)

## [1.2.2](https://github.com/cassproject/cass-vlrc/tree/1.2.2) (2019-04-15)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.2.1...1.2.2)

## [1.2.1](https://github.com/cassproject/cass-vlrc/tree/1.2.1) (2019-04-15)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.2.0...1.2.1)

**Closed issues:**

- My Goals when a different subject is selected should always be thinking about me. [\#38](https://github.com/cassproject/cass-vlrc/issues/38)
- Time block in timeline needs more height in the timeline. [\#36](https://github.com/cassproject/cass-vlrc/issues/36)

## [1.2.0](https://github.com/cassproject/cass-vlrc/tree/1.2.0) (2019-04-12)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.1.2...1.2.0)

**Closed issues:**

- New user doesn't seem to show up. [\#32](https://github.com/cassproject/cass-vlrc/issues/32)
- Switching subjects does not recalculate graph state. [\#31](https://github.com/cassproject/cass-vlrc/issues/31)
- Need to be able to delete assertions from the timeline view. [\#29](https://github.com/cassproject/cass-vlrc/issues/29)
- Remove need to click twice to add an alignment to a competency. [\#28](https://github.com/cassproject/cass-vlrc/issues/28)
- Assertions made by others in the Topic View should be under a drop down arrow \(with a number to indicate the number of positive / negative assertions\) [\#27](https://github.com/cassproject/cass-vlrc/issues/27)
- When saving assertions out to the localstorage, need to show that \<this is why this is slow\> [\#26](https://github.com/cassproject/cass-vlrc/issues/26)
- Be able to issue badge from assertion [\#21](https://github.com/cassproject/cass-vlrc/issues/21)
- Be able to attach evidence to checkmark/X [\#20](https://github.com/cassproject/cass-vlrc/issues/20)
- Need way to collapse tree nodes in list of competencies. [\#17](https://github.com/cassproject/cass-vlrc/issues/17)
- Attach ChooseActions to assertions that one knows a thing. [\#14](https://github.com/cassproject/cass-vlrc/issues/14)
- Pathway "Starring competencies in frameworks" stuff. [\#1](https://github.com/cassproject/cass-vlrc/issues/1)

## [1.1.2](https://github.com/cassproject/cass-vlrc/tree/1.1.2) (2019-03-27)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.1.1...1.1.2)

## [1.1.1](https://github.com/cassproject/cass-vlrc/tree/1.1.1) (2019-03-22)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.1.0...1.1.1)

## [1.1.0](https://github.com/cassproject/cass-vlrc/tree/1.1.0) (2019-03-15)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.0.10...1.1.0)

## [1.0.10](https://github.com/cassproject/cass-vlrc/tree/1.0.10) (2019-03-08)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.0.9...1.0.10)

**Closed issues:**

- Be able to set a competency as a goal / favorite. [\#22](https://github.com/cassproject/cass-vlrc/issues/22)

## [1.0.9](https://github.com/cassproject/cass-vlrc/tree/1.0.9) (2019-03-01)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.0.8...1.0.9)

## [1.0.8](https://github.com/cassproject/cass-vlrc/tree/1.0.8) (2019-02-21)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.0.7...1.0.8)

**Implemented enhancements:**

- Friend / invite / remove / view profile / make assertions about your friend. [\#12](https://github.com/cassproject/cass-vlrc/issues/12)

**Closed issues:**

- Need loading icon for when the list of frameworks \(or just the frameworks themselves\) are huge. [\#16](https://github.com/cassproject/cass-vlrc/issues/16)

## [1.0.7](https://github.com/cassproject/cass-vlrc/tree/1.0.7) (2019-02-15)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.0.6...1.0.7)

## [1.0.6](https://github.com/cassproject/cass-vlrc/tree/1.0.6) (2019-02-15)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.0.5...1.0.6)

## [1.0.5](https://github.com/cassproject/cass-vlrc/tree/1.0.5) (2019-02-08)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/book-slide-style...1.0.5)

## [book-slide-style](https://github.com/cassproject/cass-vlrc/tree/book-slide-style) (2019-02-05)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/0.4.26...book-slide-style)

**Closed issues:**

- Error adding assertions [\#19](https://github.com/cassproject/cass-vlrc/issues/19)

## [0.4.26](https://github.com/cassproject/cass-vlrc/tree/0.4.26) (2019-01-18)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.0.3...0.4.26)

## [1.0.3](https://github.com/cassproject/cass-vlrc/tree/1.0.3) (2019-01-17)
[Full Changelog](https://github.com/cassproject/cass-vlrc/compare/1.0.4...1.0.3)

## [1.0.4](https://github.com/cassproject/cass-vlrc/tree/1.0.4) (2019-01-17)
**Closed issues:**


js/me.js
Vue.component('profile', {
    props: ['pk', 'displayName', 'onClick'],
    data: function () {
        return {
            editing: false,
            private: false,
            personObj: null,
            refreshesStuff: true,
            inContactList: null
        }
    },
    computed: {
        person: {
            get: function () {
                if (this.personObj != null)
                    return this.personObj;
                if (this.pk == null)
                    return null;
                this.getPerson();
                return null;
            },
            set: function (newPerson) {
                this.personObj = newPerson;
            }
        },
        name: {
            get: function () {
                if (this.person == null)
                    return "Loading...";
                if (this.person.givenName != null && this.person.familyName != null)
                    return this.person.givenName +" "+ this.person.familyName;
                if (this.person.name == null)
                    return "<Restricted>";
                //EcIdentityManager.getIdentity(EcPk.fromPem(this.pk)).displayName = this.personObj.name;
                return this.person.name;
            },
            set: function (newName) {
                this.personObj.name = newName;
                if (EcIdentityManager.getIdentity(EcPk.fromPem(this.pk)) != null)
                    EcIdentityManager.getIdentity(EcPk.fromPem(this.pk)).displayName = this.personObj.name;
            }
        },
        mine: {
            get: function () {
                if (app.me == null)
                    return false;
                if (this.personObj == null)
                    return false;
                return this.personObj.hasOwner(EcPk.fromPem(app.me));
            }
        },
        fingerprint: {
            get: function () {
                if (this.personObj == null)
                    return null;
                setTimeout(function(){jdenticon()},100);
                return this.personObj.getGuid();
            }
        },
        isContact: {
            get: function () {
                if (this.inContactList == null) this.inContactList = EcIdentityManager.getContact(EcPk.fromPem(this.pk)) != null;
                return this.inContactList;
            }
        }
    },
    watch: {
        pk: function () {
            this.personObj = null;
            this.editing = false;
            this.private = false;
        },
        private: function () {
        }
    },
    methods: {
        savePerson: function () {
            this.editing = false;
            var thingToSave = this.personObj;
            var me = this;
            if (this.private) {
                if (thingToSave.reader == null) //Can delete when https://github.com/Eduworks/ec/issues/21 is resolved and integrated.
                    thingToSave.reader = []; //Can delete when https://github.com/Eduworks/ec/issues/21 is resolved and integrated.
                EcEncryptedValue.toEncryptedValueAsync(thingToSave, false, function (thingToSave) {
                    thingToSave.name = null; //Delete PII.
                    EcRepository.save(thingToSave, me.getPerson, console.error);
                }, console.error);
            } else {
                EcRepository.save(thingToSave, me.getPerson, console.error);
            }
        },
        getPerson: function () {
            this.personObj = null;
            if (this.refreshesStuff) {
                if (assertionHistory[this.pk] != null)
                    assertionHistory[this.pk].assertions = null;
                if (viewHistory[this.pk] != null)
                    viewHistory[this.pk].views = null;
            }
            var pk = EcPk.fromPem(this.pk);
            var me = this;
            EcRepository.get(repo.selectedServer + "data/" + pk.fingerprint(), function (person) {
                var e = new EcEncryptedValue();
                if (person.isAny(e.getTypes())) {
                    me.private = true;
                    e.copyFrom(person);
                    e.decryptIntoObjectAsync(function (person) {
                        var p = new Person();
                        p.copyFrom(person);
                        me.person = p;
                    }, console.error);
                } else {
                    me.private = false;
                    var p = new Person();
                    p.copyFrom(person);
                    me.person = p;
                }
            }, function (failure) {
                var pk = EcPk.fromPem(me.pk);
                var p = new Person();
                p.assignId(repo.selectedServer, pk.fingerprint());
                p.addOwner(pk);
                if (me.displayName == null)
                    p.name = "Unknown Person.";
                else
                    p.name = me.displayName;
                me.person = p;
                me.private = true;
                if (me.mine)
                    me.savePerson();
            });
        },
        cancelSave: function () {
            this.editing = false;
            this.getPerson();
        },
        clickTitle: function () {
            if (this.onClick != null)
                this.onClick(this.pk);
        },
        contact: function () {
            var c = new EcContact();
            c.pk = EcPk.fromPem(this.pk);
            c.displayName = this.name;
            EcIdentityManager.addContact(c);
            this.inContactList = true;
        },
        uncontact: function () {
            for (var i = 0; i < EcIdentityManager.contacts.length; i++) {
                if (EcIdentityManager.contacts[i].pk.toPem() == this.pk)
                    EcIdentityManager.contactChanged(EcIdentityManager.contacts.splice(i, 1));
            }
            this.inContactList = false;
        }
    },
    template: '<div v-if="person">' +
    '<span v-if="mine">' +
    '<span v-if="editing">' +
    '<i class="mdi mdi-content-save" aria-hidden="true" style="float:right;font-size:large" title="Save your person." v-on:click="savePerson()"></i>' +
    '<i class="mdi mdi-cancel" aria-hidden="true" style="float:right;font-size:large" title="Cancel editing." v-on:click="cancelSave();"></i>' +
    '</span>' +
    '<span v-else>' +
    '<i class="mdi mdi-pencil" aria-hidden="true" style="float:right;font-size:large" title="Edit your person." v-on:click="editing = true;"></i>' +
    '</span>' +
    '</span>' +
    '<span v-else>' +
    '<i class="mdi mdi-account-circle" aria-hidden="true" style="float:right;font-size:large" title="Remove person from contacts." v-if="isContact" v-on:click="uncontact();"></i>' +
    '<i class="mdi mdi-account-circle-outline" aria-hidden="true" style="float:right;font-size:large" title="Add person to contacts." v-else v-on:click="contact();"></i>' +
    '</span>' +
    '<svg style="vertical-align: sub;" width="22" height="22" v-if="fingerprint" :data-jdenticon-value="fingerprint" :title="fingerprint"></svg> <input v-if="editing" v-on:keyup.esc="cancelSave()" v-on:keyup.enter="savePerson()" v-model="name">' +
    '<h2 v-else v-on:click="clickTitle" style="display:inline;">{{ name }}</h2>' +
    '<div v-if="editing"><br><br><input :id="pk" v-model="private" type="checkbox"><label :for="pk">Private</label></div>' +
    '</div>'
});
var assertionHistory = {};
Vue.component('assertionhistory', {
    props: ['pk'],
    data: function () {
        return {
            assertionResult: [],
            searched: false
        }
    },
    computed: {
        assertions: {
            get: function () {
                var me = this;
                if (this.pk == null) return null;
                if (this.searched)
                    return this.assertionResult;
                EcAssertion.search(repo, "(\\*@reader:\"" + this.pk + "\") OR (\\*@owner:\"" + this.pk + "\")", function (assertions) {
                    assertions = assertions.sort(function (a, b) {
                        return parseInt(b.id.substring(b.id.lastIndexOf("/") + 1)) - parseInt(a.id.substring(a.id.lastIndexOf("/") + 1));
                    });
                    me.assertionResult.splice(0, me.assertionResult.length);
                    for (var i = 0; i < assertions.length; i++)
                        me.assertionResult.push(assertions[i]);
                    me.searched = true;
                }, console.error, {
                    size: 5000
                });
                return null;
            },
            set: function (assertions) {
                if (assertions == null) {
                    this.assertionResult.splice(0, this.assertionResult.length);
                }
                else {
                    this.assertionResult.splice(0, this.assertionResult.length);
                    for (var i = 0; i < assertions.length; i++)
                        this.assertionResult.push(assertions[i]);
                }
            }
        }
    },
    created: function () {
    },
    watch: {
        pk: function (newPk, oldPk) {
            delete assertionHistory[oldPk];
            this.assertionResult = [];
            this.searched = false;
            assertionHistory[newPk] = this;
        }
    },
    methods: {
        addAssertion: function (a) {
            this.assertionResult.unshift(a);
        },
        removeAssertion: function (a) {
            for (var i = 0; i < this.assertionResult.length; i++) {
                if (this.assertionResult[i].id == a.id)
                    this.assertionResult.splice(i, 1);
            }
        }
    },
    template: '<div><h3>Claims (Private)</h3>' +
    '<span v-if="assertions"><span v-if="assertions.length == 0">None.</span></span>' +
    '<ul v-if="assertions" style="max-height:10rem;overflow-y:scroll;">' +
    '<assertion v-for="item in assertions" v-bind:key="item.id" :uri="item.id"></assertion>' +
    '</ul>' +
    '<div v-else><br>Loading Assertions...</div>' +
    '</div>'
});
Vue.component('assertion', {
    props: ['uri'],
    data: function () {
        return {
            assertion: null,
            subject: null,
            agent: null,
            timestamp: null,
            expiry: null,
            competency: null,
            negative: null,
        };
    },
    computed: {
        statement: {
            get: function () {
                if (this.subject == null)
                    return "Decrypting...";
                if (this.agent == null)
                    return "Decrypting...";
                if (this.competency == null)
                    return "Decrypting...";
                var statement = "";
                if (this.timestamp != null)
                    statement += moment(this.timestamp).fromNow() + ", ";
                statement += this.agent + " claimed " + this.subject;
                if (this.negative == true)
                    statement += " could not ";
                else
                    statement += " could ";
                statement += "demonstrate " + this.competencyText + ".";
                return statement;
            }
        },
        competencyText: {
            get: function () {
                if (this.competency == null)
                    return null;
                return this.competency.name;
            }
        }
    },
    created: function () {
    },
    watch: {},
    methods: {
        initialize: function (isVisible, entry) {
            var me = this;
            if (true && isVisible) {
                EcAssertion.get(this.uri, function (assertion) {
                    me.assertion = assertion;
                    if (assertion.subject == null)
                        me.subject = "nobody"
                    else
                        assertion.getSubjectNameAsync(function (name) {
                            me.subject = name;
                        }, console.error);
                    if (assertion.agent == null)
                        me.agent = "nobody"
                    else
                        assertion.getAgentNameAsync(function (name) {
                            me.agent = name;
                        }, console.error);
                    if (assertion.assertionDate != null)
                        assertion.getAssertionDateAsync(function (assertionDate) {
                            me.timestamp = assertionDate;
                        }, console.error);
                    if (assertion.expirationDate != null)
                        assertion.getExpirationDateAsync(function (expirationDate) {
                            me.expiry = expirationDate;
                        }, console.error);
                    if (assertion.negative != null)
                        assertion.getNegativeAsync(function (negative) {
                            me.negative = negative;
                        }, console.error);
                    else
                        me.negative = false;
                    EcCompetency.get(assertion.competency, function (competency) {
                        me.competency = competency;
                    }, console.error);
                }, console.error);
            }
        }
    },
    template: '<div v-observe-visibility="{callback: initialize,once: true}"><li v-if="statement" >{{ statement }}</li></div>'

});

var viewHistory = {};
Vue.component('viewhistory', {
    props: ['pk'],
    data: function () {
        return {
            viewResult: null
        }
    },
    computed: {
        views: {
            get: function () {
                var me = this;
                if (this.pk == null) return null;
                if (this.viewResult != null)
                    return this.viewResult;
                repo.searchWithParams("@type:ChooseAction AND @owner:\"" + this.pk + "\"", {
                    size: 5000
                }, null, function (views) {
                    views = views.sort(function (a, b) {
                        return parseInt(b.id.substring(b.id.lastIndexOf("/") + 1)) - parseInt(a.id.substring(a.id.lastIndexOf("/") + 1));
                    });
                    me.viewResult = views;
                }, console.error, {
                    size: 5000
                });
                return null;
            },
            set: function (v) {
                this.viewResult = v;
            }
        }
    },
    created: function () {
        viewHistory[this.pk] = this;
    },
    watch: {
        pk: function () {
            this.viewResult = null;
        }
    },
    methods: {
        addView: function (a) {
            this.viewResult.unshift(a);
        },
        removeView: function (a) {
            for (var i = 0; i < this.viewResult.length; i++) {
                if (this.viewResult[i].id == a.id)
                    this.viewResult.splice(i, 1);
            }
        }
    },
    template: '<div><h3>Views (Public)</h3>' +
    '<span v-if="views"><span v-if="views.length == 0">None.</span></span>' +
    '<ul v-if="views" style="max-height:10rem;overflow-y:scroll;"><chooseAction v-for="item in views" v-bind:key="item.id" :uri="item.id"></chooseAction></ul>' +
    '<div v-else><br>Loading Views...</div>' +
    '</div>'
});
Vue.component('chooseAction', {
    props: ['uri'],
    data: function () {
        return {
            subject: null,
            action: null,
            resource: null
        };
    },
    computed: {
        timestamp: {
            get: function () {
                if (this.action == null) return null;
                return moment(parseInt(this.action.id.substring(this.action.id.lastIndexOf("/") + 1))).fromNow();
            }
        }
    },
    created: function () {
    },
    watch: {},
    methods: {
        initialize: function (isVisible, entry) {
            var me = this;
            if (isVisible) {
                EcRepository.get(this.uri, function (view) {
                    me.action = view;
                    EcRepository.get(repo.selectedServer + "data/schema.org.Person/" + EcPk.fromPem(view.owner[0]).fingerprint(), function (person) {
                        var e = new EcEncryptedValue();
                        if (person.isAny(e.getTypes())) {
                            e.copyFrom(person);
                            e.decryptIntoObjectAsync(function (person) {
                                var p = new Person();
                                p.copyFrom(person);
                                me.subject = p.name;
                            }, function (failure) {
                                me.subject = "someone";
                            });
                        } else {
                            var p = new Person();
                            p.copyFrom(person);
                            me.subject = p.name;
                        }
                    }, console.error);
                    EcRepository.get(view.object, function (resourceAlignment) {
                        me.resource = resourceAlignment;
                    }, console.error);
                }, console.error);
            }
        }
    },
    template: '<div v-observe-visibility="{callback: initialize,once: true}"><li v-if="subject" >{{ timestamp }}, {{ subject }} viewed <a v-if="resource" target="_blank" :href="resource.url">{{resource.name}}</a><span v-else>...</span></li></div>'

});
